### PR TITLE
fix the bug that LogicalExpressionOptimizerPass lost logical type of …

### LINKED ISSUE #64487: Incorrect optimization of multiple OR in the WHERE clause when analyzer is enabled
Tested with 24.4 in Docker.

With 3 or more variants provided in the WHERE clause with OR, a seemingly correct query returns zero results.

A minimal repro: 

```sql
SET allow_experimental_analyzer=1;
CREATE OR REPLACE TABLE foo (i Date) ENGINE MergeTree ORDER BY i;
INSERT INTO foo VALUES ('2020-01-01');
INSERT INTO foo VALUES ('2020-01-02');
```

This query returns two records, as expected:

```sql
SELECT *
FROM default.foo
WHERE (default.foo.i = parseDateTimeBestEffort('2020-01-01')) 
    OR (default.foo.i = parseDateTimeBestEffort('2020-01-02'))
ORDER BY i ASC
LIMIT 1048575
```

```
   ┌──────────i─┐
1. │ 2020-01-01 │
   └────────────┘
   ┌──────────i─┐
2. │ 2020-01-02 │
   └────────────┘
```

However, if we add a third variant via OR, the query returns an empty result set:

```sql
SELECT *
FROM default.foo
WHERE (default.foo.i = parseDateTimeBestEffort('2020-01-01'))
   OR (default.foo.i = parseDateTimeBestEffort('2020-01-02'))
   OR (default.foo.i = parseDateTimeBestEffort('2020-01-03'))
ORDER BY default.foo.i ASC
LIMIT 1048575
```

A quote from @vdimir:

> It's a bug in query tree optimization (if you are interested, check out `EXPLAIN QUERY TREE dump_ast = 1` for 3 conditions they are rewritten to `default.foo.i IN (...)` and seems this particular rewrite is incorrect). Current workaround is to disable this optimization with `SET optimize_min_equality_disjunction_chain_length = 4294967295;`

Fiddle: https://fiddle.clickhouse.com/9dd6a6a5-2496-4343-8b40-aabf57aa8bc1

### PR DESCRIPTION
**Pull request by chloro-pn**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
fix the bug that LogicalExpressionOptimizerPass lost logical type of constant. closes #64487

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by chloro-pn** [comment]
@yakov-olkhovskiy

**Comment by novikd** [comment]
Please, add a test.

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 1846705497018b01b48b7ffaea353537edfb5b67 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/66344/1846705497018b01b48b7ffaea353537edfb5b67/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Style check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/66344/1846705497018b01b48b7ffaea353537edfb5b67/style_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

**Comment by chloro-pn** [comment]
* AST fuzzer failure : #66418
* Stateless tests failure : #66419
* Stress test failure : #66026

**Comment by chloro-pn** [comment]
Other failed tests encountered an internal error in GDB:
https://github.com/ClickHouse/ClickHouse/actions/runs/9887385684/job/27358287346?pr=66344#step:8:3128
```
/build/gdb-ZgDh0V/gdb-12.1/gdb/dwarf2/expr.c:607: internal-error: coerce_pieced_ref: Assertion `closure->pieces.size () == 1' failed.
A problem internal to GDB has been detected,
further debugging may prove unreliable.
```

Do I need to restart CI ? @novikd

**Comment by chloro-pn** [comment]
* `Stateless tests (release) — fail: 1` #65589

**Comment by chloro-pn** [comment]
CI passed and can this pr be merged bro ? @novikd : )

**Review by yakov-olkhovskiy — APPROVED** [review]

File: src/Analyzer/Passes/LogicalExpressionOptimizerPass.cpp [hunk 1]
```diff
@@ -627,16 +629,18 @@ class LogicalExpressionOptimizerVisitor : public InDepthQueryTreeVisitorWithCont
...
                 {
                     const auto * lhs_literal = equals_arguments[0]->as<ConstantNode>();
                     assert(lhs_literal);
                     args.push_back(lhs_literal->getValue());
+                    tuple_element_types.push_back(lhs_literal->getResultType());
                 }
             }
 
-            auto rhs_node = std::make_shared<ConstantNode>(std::move(args));
+            auto rhs_node = std::make_shared<ConstantNode>(std::move(args), std::make_shared<DataTypeTuple>(std::move(tuple_element_types)));
```
Comment (novikd) [inline src/Analyzer/Passes/LogicalExpressionOptimizerPass.cpp:643]: Doesn't it also require adding a conversation of tuple elements to IN first argument type? I'm not sure if the semantics is correct right now.

Comment (yakov-olkhovskiy) [reply src/Analyzer/Passes/LogicalExpressionOptimizerPass.cpp:643]: I'm not quite sure what this question is about, but regarding this fix - the issue is that `DB::Field` doesn't have `DateTime` types (replacing them by some Decimal I believe), so `ConstantNode` created without explicit type argument will derive types from `Field`'s in `args`

**Comment by clickgapai** [comment]
Hi @chloro-pn — while reviewing this PR I found the following:

- **Bug (P1):** `tryOptimizeAndEqualsNotEqualsChain` loses type info when converting notEquals chain to NOT IN, causing wrong results with Date columns — https://github.com/ClickHouse/ClickHouse/issues/104537

Happy to discuss — close anything that's wrong or already addressed.

**Pull request merged**